pkg/kubelet/volume/volume_manager.go

```diff
@@ -0,0 +1,349 @@
+	uniquePodName := volumehelper.GetUniquePodName(pod)
+
+	// If pod is already mounted, force remount to reflect any changes in the
+	// referencing pod. Volumes like Downward API, depend on this to update the
+	// contents of the volume. All volume mounting calls should be idempotent so
+	// a second mount call for volumes that do not need to update contents
+	// should not fail.
+	vm.desiredStateOfWorldPopulator.ReprocessPod(uniquePodName)
+	vm.actualStateOfWorld.MarkRemountRequired(uniquePodName)
+
+	err := wait.Poll(
+		podAttachAndMountRetryInterval,
+		podAttachAndMountTimeout,
+		vm.verifyVolumesMountedFunc(uniquePodName, expectedVolumes))
+
+	if err != nil {
+		// Timeout expired
+		ummountedVolumes :=
+			vm.getUnmountedVolumes(uniquePodName, expectedVolumes)
+		if len(ummountedVolumes) == 0 {
+			return nil
+		}
+
+		return fmt.Errorf(
+			"timeout expired waiting for volumes to attach/mount for pod %q/%q. list of unattached/unmounted volumes=%v",
+			pod.Name,
+			pod.Namespace,
+			ummountedVolumes)
+	}
+
+	glog.Infof("All volumes are attached and mounted for pod %q", format.Pod(pod)) // TODO drop to V(3)?.
+	return nil
+}
+
+// verifyVolumesMountedFunc returns a method that returns true when all expected
+// volumes are mounted.
+func (vm *volumeManager) verifyVolumesMountedFunc(
+	podName types.UniquePodName, expectedVolumes []string) wait.ConditionFunc {
+	return func() (done bool, err error) {
+		return len(vm.getUnmountedVolumes(podName, expectedVolumes)) == 0, nil
+	}
+}
+
+// getUnmountedVolumes fetches the current list of mounted volumes from
+// the actual state of the world, and uses it to process the list of
+// expectedVolumes. It returns a list of unmounted volumes.
+func (vm *volumeManager) getUnmountedVolumes(
+	podName types.UniquePodName, expectedVolumes []string) []string {
+	mountedVolumes := sets.NewString()
+	for _, mountedVolume := range vm.actualStateOfWorld.GetMountedVolumesForPod(podName) {
+		mountedVolumes.Insert(mountedVolume.OuterVolumeSpecName)
+	}
+	return getUnmountedVolumesHelper(mountedVolumes, expectedVolumes)
+}
+
+// getUnmountedVolumesHelper adds each element of expectedVolumes that is not in
+// mountedVolumes to a list of unmountedVolumes and returns it.
+func getUnmountedVolumesHelper(
+	mountedVolumes sets.String, expectedVolumes []string) []string {
+	unmountedVolumes := []string{}
+	for _, expectedVolume := range expectedVolumes {
+		if !mountedVolumes.Has(expectedVolume) {
+			unmountedVolumes = append(unmountedVolumes, expectedVolume)
+		}
+	}
+	return unmountedVolumes
+}
+
+// getExpectedVolumes returns a list of volumes that should be mounted
+func getExpectedVolumes(pod *api.Pod) []string {
+	expectedVolumes := []string{}
+	if pod == nil {
+		return expectedVolumes
+	}
+
+	for _, podVolume := range pod.Spec.Volumes {
+		expectedVolumes = append(expectedVolumes, podVolume.Name)
+	}
+
+	return expectedVolumes
+}
+
+// getVolumesForPodHelper is a helper method for GetVolumesForPod methods.
```

Comment:
move to this to near the definition of `GetVolumesForPod` and `GetVolumesForPodAndAppendSupplementalGroups`